### PR TITLE
Build: spec: try finding and using libaio.so.x library name

### PR DESCRIPTION
**Pull request by gao-yan**
... i.e. the one with the major version number such as libaio.so.1, to
likely avoid requiring libaio-devel for the generic symbolic link
libaio.so.

This is achieved by adding the ability to specify a libaio name
with -DLIBAIO_NAME via CFLAGS.

This also ensures that the corresponding library name is shown in case
of dlopen() error together with more detail from dlerror()

### CODE REVIEW

**Comment by wenningerk** [comment]
Ooops obviously a leftover.
Anyway did you check what dlerror() would give us here?
Man page doesn't tell if it is a simple error-code translation or if we get an actual reference to the library it failed on.
Maybe more descriptive and we might not have to keep that in sync.

**Comment by gao-yan** [comment]
Well, this is only about the inconsistency...

Actually the main issue for me now is, at least in the seemingly convention of SUSE Linux, library files always have a major version as the suffix, for instance `libaio.so.1` is a symbolic link to `libaio.so.1.0.1` while there's not a `libaio.so`.

So sbd test fails to find the library file at all for me now.

**Comment by wenningerk** [comment]
That is the reason why it was initially libaio.so.1 - as left over in the logs.
Don't remember the reason for the switch. I remember there was something in tumbleweed iirc but I'm not sure that this was the reason to switch. Let me consult the git-logs - maybe that rings a bell ...

**Comment by gao-yan** [comment]
Makes sense.

BTW, need to look more into it, but probably depending on the versions of the toolchain (glibc/bintuils...) , it seems to work in certain cases without requiring the specific file  libaiso.so to be existing ...

**Comment by wenningerk** [comment]
Maybe some kind of automatism in the dynamic library loader ...
If you have explicit cases that work and some that don't it would be helpful to have a hint for getting that covered in CI if somehow possible.

**Comment by gao-yan** [comment]
Alright, I see. It works only if libaio-devel is as well installed, which provides the symbolic link `libaio.so`.

**Comment by wenningerk** [comment]
Helpful finding! Possibility would be to include the devel requirement into the test-subpackage. But that would probably not be too clean. Best might be to have configure find what libaio.so is linking to and bake that into the test-subpackage.

**Comment by gao-yan** [comment]
> Helpful finding! (That brings back my memory that I had switched to libaio.so for that version of tumbleweed that came with 2 versions of libaio.)

Hrm, strangely I can only see a single version libaio1-0.3.113 in tumbleweed.

> Possibility would be to include the devel requirement into the test-subpackage. But that would probably not be too clean.  

Not a bad solution before we have a perfect one.

> Best might be to have configure find what libaio.so is linking to and bake that into the test-subpackage.

AFAICT, it's not straightforward where exactly `AC_CHECK_LIB` or basically `gcc -laio` may find the library...

Something like `ldconfig -p` could do but ldconfig usually resides in /sbin,  which seems to mean it's supposed be for the privileged user.

 It would be helpful if libaio-devel shipped a pkg-config file but unfortunately not ...

File: sbd.spec [hunk 1]
```diff
@@ -97,6 +97,7 @@ Available rpmbuild rebuild options:
 Summary:        Storage-based death environment for regression tests
 License:        GPL-2.0-or-later
 Group:          System Environment/Daemons
+Requires:       libaio-devel
```
Comment (wenningerk) [inline sbd.spec:100]: The reason for the requirement here is not obvious. So we might want to have a comment to first remember why we did it and to stumble over it to find a better solution (best before the next release).

**Comment by wenningerk** [comment]
I'm afraid autotools don't have anything available to get the actual library name. something like  ```readlink -f %{_libdir}libaio.so|xargs basename``` could be used to keep the ugliness at least in the spec-file and avoid that the test-packages pulls in libaio-devel.

**Comment by gao-yan** [comment]
Thing is we are not sure whether libaio.so really necessarily resides in a standard {_libidr}, right? Reason why I was forward to a pkg-config from libaio-devel ...

**Comment by wenningerk** [comment]
Got why you'd like to have the pkg-config. But not having pkg-config wouldn't it be very unlikely not to have %{_libaio.so}? For an rpm-based distribution I would at least strongly assume it should be there. Which is why I was suggesting to do that in the spec-file. And if it is still not found we could still fall back to "libaio.so" (and maybe Requiring libaio-devel) as we would if the macro isn't set via  CFLAGS.
Well there is something ugly going on that requires us to explicitly give the library. And everything we're doing to live with it seems to be somehow ugly as well.

**Comment by gao-yan** [comment]
We could surely try harder to likely avoid requiring libaio-devel.

> +export CFLAGS="$RPM_OPT_FLAGS -Wall -Werror -DLIBAIO_LIB=$(readlink -f %{_libdir}[libaio.so](http://libaio.so/)|xargs basename)"

This would make it unnecessarily less compatible to strictly bind to the very specific library name with the full version numbers. The one with the major version number should be ideal.

```
> +#define LIBAIO_LIB_DOQUOTE(lib) #lib
> +#ifndef LIBAIO_LIB
> +  #define LIBAIO_LIB_QUOTED LIBAIO_LIB_DOQUOTE(libaio.so)
> +#else
> +  #define LIBAIO_LIB_QUOTED LIBAIO_LIB_DOQUOTE(LIBAIO_LIB)
> +#endif
```
This would quote and use the macro name itself i.e. `"LIBAIO_LIB"`.

We have to use two levels of macros to stringize the result of macro expansion:
https://gcc.gnu.org/onlinedocs/cpp/Stringizing.html

```
> +        handle = dlopen(LIBAIO_LIB_QUOTED,  RTLD_NOW);
```
We'd better still check and avoid passing a possibly empty string here.

**Comment by wenningerk** [comment]
Well said I didn't try ;-)

Below I did test this time. Reduction to the major would've made the tumbleweed fail which was why I had gone from hardcoded 'libaio.so.1' to 'libaio.so'. But I guess the repo was probably broken/inconsistent and should have failed anyway.
Why would you btw. check the library-name once again? When sbdy defines LIBAIO_LIB to something we should be able to assume it is something useful otherwise we would default to 'libaio.so' anyway.

```
diff --git a/sbd.spec b/sbd.spec
index 0291598..8bba17b 100644
--- a/sbd.spec
+++ b/sbd.spec
@@ -109,7 +109,7 @@ regression-testing sbd.
 
 %build
 ./autogen.sh
-export CFLAGS="$RPM_OPT_FLAGS -Wall -Werror"
+export CFLAGS="$RPM_OPT_FLAGS -Wall -Werror -DLIBAIO_LIB=$(readlink -f %{_libdir}/libaio.so|xargs basename|cut -d "." -f1-3)"
 %configure --with-watchdog-timeout-default=%{watchdog_timeout_default} \
            --with-sync-resource-startup-default=%{?with_sync_resource_startup_default:yes}%{!?with_sync_resource_startup_default:no}  \
            --with-sync-resource-startup-sysconfig=%{sync_resource_startup_sysconfig} \
diff --git a/tests/sbd-testbed.c b/tests/sbd-testbed.c
index 02844a7..9e86e24 100644
--- a/tests/sbd-testbed.c
+++ b/tests/sbd-testbed.c
@@ -56,6 +56,14 @@ g_unix_fd_add(gint fd,
 }
 #endif
 
+#define LIBAIO_LIB_DOQUOTE(lib) xLIBAIO_LIB_DOQUOTE(lib)
+#define xLIBAIO_LIB_DOQUOTE(lib) #lib
+#ifndef LIBAIO_LIB
+  #define LIBAIO_LIB_QUOTED LIBAIO_LIB_DOQUOTE(libaio.so)
+#else
+  #define LIBAIO_LIB_QUOTED LIBAIO_LIB_DOQUOTE(LIBAIO_LIB)
+#endif
+
 typedef int (*orig_open_f_type)(const char *pathname, int flags, ...);
 typedef int (*orig_ioctl_f_type)(int fd, unsigned long int request, ...);
 typedef ssize_t (*orig_write_f_type)(int fd, const void *buf, size_t count);
@@ -154,9 +162,9 @@ init (void)
         orig_fopen        = (orig_fopen_f_type)dlsym_fatal(RTLD_NEXT,"fopen");
         orig_fclose       = (orig_fclose_f_type)dlsym_fatal(RTLD_NEXT,"fclose");
 
-        handle = dlopen("libaio.so",  RTLD_NOW);
+        handle = dlopen(LIBAIO_LIB_QUOTED,  RTLD_NOW);
         if (!handle) {
-            fprintf(stderr, "Failed opening libaio.so.1\n");
+            fprintf(stderr, "Failed opening %s(%s)\n", LIBAIO_LIB_QUOTED, dlerror());
             exit(1);
         }
         orig_io_setup     = (orig_io_setup_f_type)dlsym_fatal(handle,"io_setup");
```

File: tests/sbd-testbed.c [hunk 2]
```diff
@@ -154,9 +166,13 @@ init (void)
         orig_fopen        = (orig_fopen_f_type)dlsym_fatal(RTLD_NEXT,"fopen");
         orig_fclose       = (orig_fclose_f_type)dlsym_fatal(RTLD_NEXT,"fclose");
 
-        handle = dlopen("libaio.so",  RTLD_NOW);
+        if (libaio == NULL || libaio[0] == '\0') {
+            libaio = LIBAIO_DEFAULT;
```
Comment (wenningerk) [inline tests/sbd-testbed.c:170]: wouldn't we want to use LIBAIO_QUOTED here then.

**Comment by wenningerk** [comment]
Not sure if that complies to any rules but iirc I've seen bogus stuff at the end of the versioning.
I think it was even I changed to ```libaio.so``` from hardcoded ```libaio.so.1```.
But reduction to the major would as well not have removed the crash back then - well probably the repo was just broken/inconsistent and should have failed.
Wouldn't it be safer to start at the front like ```cut -d "." -f1-3```?

**Comment by gao-yan** [comment]
> Wouldn't it be safer to start at the front like `cut -d "." -f1-3`?

Good idea. Should be safer this way.

**Comment by wenningerk** [comment]
> > Wouldn't it be safer to start at the front like `cut -d "." -f1-3`?
> 
> Good idea. Should be safer this way.

At least more towards the intention ... for the case mentioned with tumbleweed back then your approach probably would even have the test made work - no match - no replacement - go for the full version. But as said maybe it shouldn't have.

**Comment by wenningerk** [comment]
Thanks!

**Comment by gao-yan** [comment]
Thanks for your good ideas!

**Pull request merged**